docs/assets.rst

```diff
@@ -0,0 +1,673 @@
+``CachedFileFinder`` strips hashes from filenames to identify the
+"pre-cached" names for files, by removing the middle element of filenames
+with three dots. This may have been useful in django-pipeline 1.3 or earlier,
+but it appears to do nothing now, or could potentially do the wrong thing
+such as resolving ``bootstrap.min.js`` as ``bootstrap.js``.
+
+.. _pipeline.finders.PipelineFinder:
+
+``PipelineFinder`` does nothing if ``PIPELINE['PIPELINE_ENABLED']`` if
+``True`` (testing and production), and uses the Storage to find files if it
+is disabled. For Kuma, this means it may find files in the ``STATIC_ROOT``
+directory. However, since the FileSystemFinder_ finds most files in
+``kuma/static`` first, it is doubtful if this Finder ever applies.
+
+Storage
+*******
+
+Most of the functionality of ``django-pipeline`` is implemented as a
+:ref:`Storage class <staticfiles-storage>`, and Kuma uses three different
+implementations depending on the environment.
+
+.. _pipeline.storage.PipelineStorage:
+
+The simplest storage, used during testing and in the ``Makefile``, is
+``pipeline.storage.PipelineStorage``, which extends
+the :ref:`staticfiles Storage class <staticfiles-storage>`
+``StaticFilesStorage``, with a ``post_process`` step that packages JS and CSS
+into one-file bundles, according to the ``PIPELINE`` configuration.
+
+.. _pipeline.storage.NonPackagingPipelineStorage:
+
+Development uses ``pipeline.storage.NonPackagingPipelineStorage``.
+This works the same way as ``PipelineStorage``, but avoids creating packages,
+where several files are combined into one. JavaScript files are
+served from the source folders, but CSS files need to be compiled from Sass_,
+and are served from the ``/static`` folder after collection. When developing
+style files, a developer either needs to run ``./manage.py collectstatic`` or
+use :ref:`gulp <gulp-alternative>` to see changes.
+
+.. _kuma.core.pipeline.storage.ManifestPipelineStorage:
+
+In production, ``kuma.core.pipeline.storage.ManifestPipelineStorage`` is used.
+This combines the package processing of ``PipelineStorage`` with the hashed
+assets and ``staticfiles.json`` of ``ManifestStaticFilesStorage``. These are
+generated when the production Docker containers are created.
+
+.. _`django-pipeline library`: https://django-pipeline.readthedocs.io/en/latest/
+.. _node-sass: https://github.com/sass/node-sass
+.. _cleancss: https://github.com/jakubpawlowicz/clean-css-cli
+.. _`configuration items`: https://django-pipeline.readthedocs.io/en/latest/configuration.html
+.. _SASS: https://sass-lang.com/
+
+.. _gulp-alternative:
+
+Compiling and collecting assets with Gulp
+=========================================
+An alternate way to compile and collect assets is to use Gulp, as described in
+:ref:`compiling-with-gulp`. This requires installing node and related packages
+on the "host" system, rather than relying on the Docker containers, but it
+matches the preferred workflow of some front-end developers.
```

Comment:
I'm not sure I every really appreciated that gulp is a front-end alternative to collectstatic. So if I'm understanding correctly, we've got different build workflows for development, testing, and production, and also, frontend and backend developers have different versions of the development flow.

There is probably some kind of diagram that could summarize this. I'm not suggesting that you take any time to create such a diagram, however.

Reply:
I've tried to draw it three times 😁 Maybe the fourth will look sane...

Reply:
Nope, failed the fourth time as well. I'm going to concentrate on removing ``django-pipeline`` so there is no more variety or magic.